rhino/src/main/java/org/mozilla/javascript/ScriptableObject.java

```diff
@@ -1729,30 +1744,47 @@ public void defineProperty(
         if (getter == null && setter == null)
             throw ScriptRuntime.typeError("at least one of {getter, setter} is required");
 
+        LambdaAccessorSlot newSlot = createLambdaAccessorSlot(name, 0, getter, setter, attributes);
+        ScriptableObject newDesc = newSlot.buildPropertyDescriptor(cx);
+        checkPropertyDefinition(newDesc);
         slotMap.compute(
                 name,
                 0,
-                (id, index, existing) ->
-                        ensureLambdaAccessorSlot(
-                                cx, id, index, existing, getter, setter, attributes));
+                (id, index, existing) -> {
+                    if (existing != null) {
+                        // it's dangerous to use `this` as scope inside slotMap.compute.
+                        // It can cause deadlock when ThreadSafeSlotMapContainer is used
+
+                        return replaceExistingLambdaSlot(cx, name, existing, newSlot);
```

Comment:
Just looking at code coverage, since you're changing some fundamental stuff here, and I see that line 1758 doesn't get covered. Do you think that you could devise a test case for that and add it to your new test?

Reply:
Thanks for pointing this out, [474c3ba](https://github.com/mozilla/rhino/pull/1746/commits/474c3ba6ad89d2067dba48ec98d0b98b2e959a3d) should take care of it.